pkg/kubelet/kubelet_network_test.go

```diff
@@ -29,38 +29,88 @@ import (
 )
 
 func TestNodeIPParam(t *testing.T) {
-	testKubelet := newTestKubelet(t, false /* controllerAttachDetachEnabled */)
-	defer testKubelet.Cleanup()
-	kubelet := testKubelet.kubelet
-	tests := []struct {
+	type test struct {
 		nodeIP   string
 		success  bool
 		testName string
-	}{
+	}
+	tests := []test{
 		{
 			nodeIP:   "",
-			success:  true,
+			success:  false,
 			testName: "IP not set",
 		},
 		{
 			nodeIP:   "127.0.0.1",
 			success:  false,
-			testName: "loopback address",
+			testName: "IPv4 loopback address",
+		},
+		{
+			nodeIP:   "::1",
+			success:  false,
+			testName: "IPv6 loopback address",
+		},
+		{
+			nodeIP:   "224.0.0.1",
+			success:  false,
+			testName: "multicast IPv4 address",
+		},
+		{
+			nodeIP:   "ff00::1",
+			success:  false,
+			testName: "multicast IPv6 address",
+		},
+		{
+			nodeIP:   "169.254.0.1",
+			success:  false,
+			testName: "IPv4 link-local unicast address",
+		},
+		{
+			nodeIP:   "fe80::0202:b3ff:fe1e:8329",
+			success:  false,
+			testName: "IPv6 link-local unicast address",
+		},
+		{
+			nodeIP:   "0.0.0.0",
+			success:  false,
+			testName: "Unspecified IPv4 address",
 		},
 		{
-			nodeIP:   "FE80::0202:B3FF:FE1E:8329",
+			nodeIP:   "::",
 			success:  false,
-			testName: "IPv6 address",
+			testName: "Unspecified IPv6 address",
 		},
 		{
 			nodeIP:   "1.2.3.4",
 			success:  false,
 			testName: "IPv4 address that doesn't belong to host",
 		},
 	}
+	addrs, err := net.InterfaceAddrs()
```

Comment:
Looks like your taking some live system addresses, to provide more test cases. It would be better, IMHO, to just add static test cases with valid V4 and V6 addresses. It removes variability in the tests, makes it easier to see what is tested (easily seen in tests struct), is faster, has less duplication (if system has many interfaces), and is more in the spirit of unit testing.

Just add a valid IPv4 and a valid IPv6 test case.

Reply:
ValidateNodeIP not only checks that it's a valid IP, but also ensures the IP is on the system.

Reply:
OK, after discussing with you in person, I see you are also checking that the IP is on the system. As a future cleanup, we could split out the validation of the IP against the system's IPs into a separate function, and then UT that, without calling InterfaceAddrs (and instead pass in fake addresses). So this UT could test just failing cases, and another UT could test the function that compares the IP to system IPs.